File: reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = jump = ListNode(0) # Create a temp head node
        dummy.next = left = right = head

        while True: # Run through the list
            count = 0
            while right and count < k: # Check to make sure reverse is possible
                right = right.next
                count += 1
            if count == k:
                prev, current = right, left
                for _ in range(k):
                    current.next, current, prev = prev, current.next, current # Reverse
                jump.next, jump, left = prev, left, right # Reconnect
            else:
                return dummy.next

File: reverse-nodes-in-k-group/reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution 
{
    public ListNode reverseKGroup(ListNode head, int k) 
    {
        // Find the length of the linked list
        int length = 0;
        for (ListNode node = head; node != null; length++, node = node.next);
        
        // Iterate the linked list while there are k nodes to reverse
        ListNode dummy = new ListNode(-1, head);
        for (ListNode prev = dummy, tail = head; length >= k; length -= k)
        {
            // Reverse the k nodes
            for (int i = 1; i < k; i++)
            {
                ListNode current = tail.next.next;
                tail.next.next = prev.next;
                prev.next = tail.next;
                tail.next = current;
            }
            // Update pointers for another iteration
            prev = tail;
            tail = tail.next;
        }
        
        return dummy.next;
    }
}